File: array/best-time-to-buy-and-sell-stock.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
    return 0 if prices.empty?

    best_buy = prices[0]
    profit = 0
    prices.each do |price|
        profit = [profit, price - best_buy].max
        best_buy = [best_buy, price].min
    end
    profit
end

File: array/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res = 0
        best_buy = prices[0]
        for i in range(len(prices)):
            profit = prices[i] - best_buy
            res = max(res, profit)
            best_buy = min(best_buy, prices[i])
        return res

"""
[7,1,5,3,6,4]
 ^
   ^
"""